Roborio running out of memory

Our roboRIO is throwing this error: 

```
Loading Dependency Injection Framework
Feb 27, 2015 2:09:45 AM java.util.logging.LogManager$RootLogger log
CONFIG: Configuration done.
Feb 27, 2015 2:09:45 AM java.util.logging.LogManager$RootLogger log
CONFIG: GRIP Version: 1.1.1
platform: /Linux/arm/
Feb 27, 2015 2:09:50 AM edu.wpi.grip.core.Main start
INFO: Loading file /home/lvuser/project.grip
Java HotSpot(TM) Embedded Client VM warning: INFO: os::commit_memory(0xb48c2000, 32768, 1) failed; error='Cannot allocate memory' (errno=12)

 There is insufficient memory for the Java Runtime Environment to continue.
 Native memory allocation (mmap) failed to map 32768 bytes for committing reserved memory.
 An error report file with more information is saved as:
 /home/lvuser/hs_err_pid1963.log

 Compiler replay data is saved as:
 /home/lvuser/replay_pid1963.log
```

Anyone else run into this problem? Or any ideas for fixes?

Here are a couple more examples of out of memory.  I may run the robot program 3 to 10 times without trouble then the next 1 to 4 times in a row get the out of memory.  Rebooting and restarting don't seem to have much effect in predicting when the error will happen.

There is only 1 java process running project.grip - the ps command display of processes appears okay.

Do you have a way of trying one or more of the suggestions in the error log to prevent this error?

ONE OF THE JAVA ERROR LOGS:

roboRIO-4237-FRC:/tmp$ cat hs_err_pid9636.log

 There is insufficient memory for the Java Runtime Environment to continue.
 Native memory allocation (mmap) failed to map 1048576 bytes for committing reserved memory.
 Possible reasons:
   The system is out of physical RAM or swap space
   In 32 bit mode, the process size limit was hit
 Possible solutions:
   Reduce memory load on the system
   Increase physical memory or swap space
   Check if swap backing store is full
   Use 64 bit Java on a 64 bit OS
   Decrease Java heap size (-Xmx/-Xms)
   Decrease number of Java threads
   Decrease Java thread stack sizes (-Xss)
   Set larger code cache with -XX:ReservedCodeCacheSize=
 This output file may be truncated or incomplete.

  Out of Memory Error (os_linux.cpp:2718), pid=9636, tid=3027186784

 JRE version: Java(TM) SE Embedded Runtime Environment (8.0_06-b23) (build 1.8.0_06-b23)
 Java VM: Java HotSpot(TM) Embedded Client VM (25.6-b23 mixed mode linux-arm )
 Core dump written. Default location: //core or core.9636 (max size 2048 kB). To ensure a full core dump, try "ulimit -c unlimited" before starting Java again

========================================
âž” Launching Â«'/home/lvuser/FRCUserProgram'Â»
[HAL] Starting task FRC_DriverStation...
[GRIP Test] in Constructor
[GRIP Test] in RobotInit
[GRIP Test] fork okay continuing with parent robot code
Default Disabled() method... Override me!
NT: ERROR: bind() failed: Address already in use (TCPAcceptor.cpp:102)
Loading Dependency Injection Framework
Jan 24, 2016 4:46:22 PM java.util.logging.LogManager$RootLogger log
CONFIG: Configuration done.
Jan 24, 2016 4:46:22 PM java.util.logging.LogManager$RootLogger log
CONFIG: GRIP Version: 1.1.1
Java HotSpot(TM) Embedded Client VM warning: INFO: os::commit_memory(0xae53a000, 131072, 0) failed; error='Cannot allocate memory' (errno=12)

 There is insufficient memory for the Java Runtime Environment to continue.
 Native memory allocation (mmap) failed to map 131072 bytes for committing reserved memory.
 An error report file with more information is saved as:
 /tmp/hs_err_pid13039.log

=======


? Launching «'/home/lvuser/FRCUserProgram'»
[HAL] Starting task FRC_DriverStation...
GRIP Test in RobotInit
fork okay back in parent robot code
Default DisabledInit() method... Overload me!
Default DisabledPeriodic() method... Overload me!
NT: server: client CONNECTED: 169.254.248.235 port 51495
NT: server: client CONNECTED: 169.254.6.209 port 59054
NT: server: client CONNECTED: 169.254.6.209 port 59055
Loading Dependency Injection Framework
Jan 24, 2016 3:04:43 PM java.util.logging.LogManager$RootLogger log
CONFIG: Configuration done.
Jan 24, 2016 3:04:43 PM java.util.logging.LogManager$RootLogger log
CONFIG: GRIP Version: 1.1.1
platform: /Linux/arm/
Jan 24, 2016 3:04:51 PM edu.wpi.grip.core.Main start
INFO: Loading file /home/lvuser/project.grip
Jan 24, 2016 3:04:55 PM edu.wpi.grip.core.GRIPCoreModule onThreadException
SEVERE: Error from Thread[main,5,main]
java.lang.UnsatisfiedLinkError: java.io.IOException: No space left on device
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:613)
	at org.bytedeco.javacpp.Loader.load(Loader.java:438)
	at org.bytedeco.javacpp.Loader.load(Loader.java:381)
	at org.bytedeco.javacpp.opencv_imgcodecs.<clinit>(opencv_imgcodecs.java:13)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:340)
	at org.bytedeco.javacpp.Loader.load(Loader.java:413)
	at org.bytedeco.javacpp.Loader.load(Loader.java:381)
	at org.bytedeco.javacpp.opencv_videoio.<clinit>(opencv_videoio.java:14)
	at org.bytedeco.javacv.OpenCVFrameGrabber.start(OpenCVFrameGrabber.java:177)
	at org.bytedeco.javacv.FrameGrabber.restart(FrameGrabber.java:430)
	at edu.wpi.grip.core.sources.CameraSource.start(CameraSource.java:189)
	at edu.wpi.grip.core.sources.CameraSource.initialize(CameraSource.java:174)
	at edu.wpi.grip.core.Source.initializeSafely(Source.java:98)
	at edu.wpi.grip.core.serialization.SourceConverter.unmarshal(SourceConverter.java:60)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readItem(AbstractCollectionConverter.java:71)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.addCurrentElementToCollection(CollectionConverter.java:98)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:91)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:85)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.unmarshal(CollectionConverter.java:80)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.start(TreeUnmarshaller.java:134)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.unmarshal(AbstractTreeMarshallingStrategy.java:32)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1206)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1190)
	at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1061)
	at edu.wpi.grip.core.serialization.Project.open(Project.java:68)
	at edu.wpi.grip.core.serialization.Project.open(Project.java:62)
	at edu.wpi.grip.core.Main.start(Main.java:59)
	at edu.wpi.grip.core.Main.main(Main.java:41)
Caused by: java.io.IOException: No space left on device
	at java.io.FileOutputStream.writeBytes(Native Method)
	at java.io.FileOutputStream.write(FileOutputStream.java:307)
	at org.bytedeco.javacpp.Loader.extractResource(Loader.java:306)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:563)
	... 39 more

Java HotSpot(TM) Embedded Client VM warning: INFO: os::commit_memory(0xb68aa000, 12288, 0) failed; error='Cannot allocate memory' (errno=12)

 There is insufficient memory for the Java Runtime Environment to continue.
 Native memory allocation (mmap) failed to map 12288 bytes for committing reserved memory.
 An error report file with more information is saved as:
 /tmp/hs_err_pid3388.log
=====
The project.grip:
< grip:Pipeline>
  < sources>
    < grip:Camera>
      < property name="deviceNumber" value="0"/>
    < /grip:Camera>
  < /sources>
  < steps>
    < grip:Step name="HSL Threshold">
      < grip:Input step="0" socket="0"/>
      < grip:Input step="0" socket="1">
        < value>
          < double>0.0< /double>
          < double>140.06825938566553< /double>
        < /value>
      < /grip:Input>
      < grip:Input step="0" socket="2">
        < value>
          < double>123.83093525179855< /double>
          < double>148.38737201365188< /double>
        < /value>
      < /grip:Input>
      < grip:Input step="0" socket="3">
        < value>
          < double>9.172661870503596< /double>
          < double>255.0< /double>
        < /value>
      < /grip:Input>
      < grip:Output step="0" socket="0" previewed="false"/>
    < /grip:Step>
    < grip:Step name="Find Contours">
      < grip:Input step="1" socket="0"/>
      < grip:Input step="1" socket="1">
        < value>true< /value>
      < /grip:Input>
      < grip:Output step="1" socket="0" previewed="false"/>
    < /grip:Step>
    < grip:Step name="Filter Contours">
      < grip:Input step="2" socket="0"/>
      < grip:Input step="2" socket="1">
        < value>10.0< /value>
      < /grip:Input>
      < grip:Input step="2" socket="2">
        < value>0.0< /value>
      < /grip:Input>
      < grip:Input step="2" socket="3">
        < value>2.0< /value>
      < /grip:Input>
      < grip:Input step="2" socket="4">
        < value>1000.0< /value>
      < /grip:Input>
      < grip:Input step="2" socket="5">
        < value>0.0< /value>
      < /grip:Input>
      < grip:Input step="2" socket="6">
        < value>1000.0< /value>
      < /grip:Input>
      < grip:Input step="2" socket="7">
        < value>
          < int>0< /int>
          < int>100< /int>
        < /value>
      < /grip:Input>
      < grip:Output step="2" socket="0" previewed="false"/>
    < /grip:Step>
    < grip:Step name="Publish ContoursReport">
      < grip:Input step="3" socket="0"/>
      < grip:Input step="3" socket="1">
        < value>MyTargetTable< /value>
      < /grip:Input>
      < grip:Input step="3" socket="2">
        < value>false< /value>
      < /grip:Input>
      < grip:Input step="3" socket="3">
        < value>false< /value>
      < /grip:Input>
      < grip:Input step="3" socket="4">
        < value>false< /value>
      < /grip:Input>
      < grip:Input step="3" socket="5">
        < value>true< /value>
      < /grip:Input>
      < grip:Input step="3" socket="6">
        < value>true< /value>
      < /grip:Input>
      < grip:Input step="3" socket="7">
        < value>false< /value>
      < /grip:Input>
    < /grip:Step>
    < grip:Step name="Find Blobs">
      < grip:Input step="4" socket="0"/>
      < grip:Input step="4" socket="1">
        < value>6.0< /value>
      < /grip:Input>
      < grip:Input step="4" socket="2">
        < value>
          < double>0.0< /double>
          < double>0.8464163822525597< /double>
        < /value>
      < /grip:Input>
      < grip:Input step="4" socket="3">
        < value>true< /value>
      < /grip:Input>
      < grip:Output step="4" socket="0" previewed="false"/>
    < /grip:Step>
    < grip:Step name="Publish BlobsReport">
      < grip:Input step="5" socket="0"/>
      < grip:Input step="5" socket="1">
        < value>myBlobsReport< /value>
      < /grip:Input>
      < grip:Input step="5" socket="2">
        < value>true< /value>
      < /grip:Input>
      < grip:Input step="5" socket="3">
        < value>true< /value>
      < /grip:Input>
      < grip:Input step="5" socket="4">
        < value>true< /value>
      < /grip:Input>
    < /grip:Step>
  < /steps>
  < connections>
    < grip:Connection>
      < grip:Output source="0" socket="0" previewed="false"/>
      < grip:Input step="4" socket="0"/>
    < /grip:Connection>
    < grip:Connection>
      < grip:Output step="4" socket="0" previewed="false"/>
      < grip:Input step="5" socket="0"/>
    < /grip:Connection>
    < grip:Connection>
      < grip:Output source="0" socket="0" previewed="false"/>
      < grip:Input step="0" socket="0"/>
    < /grip:Connection>
    < grip:Connection>
      < grip:Output step="0" socket="0" previewed="false"/>
      < grip:Input step="1" socket="0"/>
    < /grip:Connection>
    < grip:Connection>
      < grip:Output step="2" socket="0" previewed="false"/>
      < grip:Input step="3" socket="0"/>
    < /grip:Connection>
    < grip:Connection>
      < grip:Output step="1" socket="0" previewed="false"/>
      < grip:Input step="2" socket="0"/>
    < /grip:Connection>
  < /connections>
  < settings>
    < teamNumber>4237< /teamNumber>
    < publishAddress>roborio-4237-frc.local< /publishAddress>
    < deployAddress>roborio-4237-frc.local< /deployAddress>
    < deployDir>/home/lvuser< /deployDir>
    < deployUser>lvuser< /deployUser>
    < deployJavaHome>/usr/local/frc/JRE/< /deployJavaHome>
  < /settings>
< /grip:Pipeline>roboRIO-4237-FRC:~$

--- a/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
+++ b/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
@@ -31,6 +31,9 @@ public class ProjectSettings implements Cloneable {
     @Setting(label = "Deploy Java home", description = "Where Java is installed on the robot.")
     private String deployJavaHome = "/usr/local/frc/JRE/";
 
+    @Setting(label = "Deploy JVM options", description = "Command line options passed to the roboRIO JVM")
+    private String deployJvmOptions = "-Xmx50m";
+
     /**
      * Set the FRC team number.  If the deploy address and NetworkTables server address haven't been manually
      * overridden, this also changes them to the mDNS hostname of the team's roboRIO.
@@ -98,6 +101,14 @@ public class ProjectSettings implements Cloneable {
         if (deployJavaHome != null) this.deployJavaHome = deployJavaHome;
     }
 
+    public String getDeployJvmOptions() {
+        return deployJvmOptions;
+    }
+
+    public void setDeployJvmOptions(String deployJvmOptions) {
+        this.deployJvmOptions = deployJvmOptions;
+    }
+
     private String computeFRCAddress(int teamNumber) {
         return "roborio-" + teamNumber + "-frc.local";
     }
@@ -109,6 +120,7 @@ public class ProjectSettings implements Cloneable {
                 .add("deployDir", deployDir)
                 .add("deployUser", deployUser)
                 .add("deployJavaHome", deployJavaHome)
+                .add("deployJvmOptions", deployJvmOptions)
                 .add("publishAddress", publishAddress)
                 .add("teamNumber", teamNumber)
                 .toString();

--- a/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
+++ b/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
@@ -106,7 +106,7 @@ public class ProjectSettings implements Cloneable {
     }
 
     public void setDeployJvmOptions(String deployJvmOptions) {
-        this.deployJvmOptions = deployJvmOptions;
+        if (deployJvmOptions != null) this.deployJvmOptions = deployJvmOptions;
     }
 
     private String computeFRCAddress(int teamNumber) {

--- a/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
+++ b/core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java
@@ -32,7 +32,7 @@ public class ProjectSettings implements Cloneable {
     private String deployJavaHome = "/usr/local/frc/JRE/";
 
     @Setting(label = "Deploy JVM options", description = "Command line options passed to the roboRIO JVM")
-    private String deployJvmOptions = "-Xmx50m";
+    private String deployJvmOptions = "-Xmx50m -XX:-OmitStackTraceInFastThrow -XX:+HeapDumpOnOutOfMemoryError";
 
     /**
      * Set the FRC team number.  If the deploy address and NetworkTables server address haven't been manually

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -12,7 +12,6 @@ import edu.wpi.grip.ui.util.StringInMemoryFile;
 import javafx.application.Platform;
 import javafx.beans.binding.Bindings;
 import javafx.beans.property.BooleanProperty;
-import javafx.beans.property.SimpleBooleanProperty;
 import javafx.fxml.FXML;
 import javafx.scene.control.*;
 import net.schmizz.sshj.SSHClient;
@@ -55,22 +54,19 @@ public class DeployController {
     @FXML private TextField deployDir;
     @FXML private ProgressIndicator progress;
     @FXML private Label command;
-    @FXML private Button deployButton;
     @FXML private Label status;
     @FXML private TextArea console;
+    @FXML private BooleanProperty deploying;
 
     @Inject private EventBus eventBus;
     @Inject private Project project;
     @Inject private Pipeline pipeline;
-
     @Inject private Logger logger;
-    private final BooleanProperty deploying = new SimpleBooleanProperty(this, "deploying", false);
+
     private Optional<Thread> deployThread = Optional.empty();
 
     @FXML
     public void initialize() {
-        deployButton.disableProperty().bind(deploying);
-        progress.disableProperty().bind(deploying.not());
         deploying.addListener((o, b, d) -> progress.setProgress(d ? ProgressIndicator.INDETERMINATE_PROGRESS : 0));
         command.textProperty().bind(Bindings.concat(javaHome.textProperty(), "/bin/java ", jvmArgs.textProperty(),
                 " -jar '", deployDir.textProperty(), "/", GRIP_JAR, "' '", deployDir.textProperty(), "/", PROJECT_FILE, "'"));

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -10,6 +10,7 @@ import edu.wpi.grip.core.serialization.Project;
 import edu.wpi.grip.core.settings.ProjectSettings;
 import edu.wpi.grip.ui.util.StringInMemoryFile;
 import javafx.application.Platform;
+import javafx.beans.binding.Bindings;
 import javafx.beans.property.BooleanProperty;
 import javafx.beans.property.SimpleBooleanProperty;
 import javafx.fxml.FXML;
@@ -50,8 +51,10 @@ public class DeployController {
     @FXML private TextField user;
     @FXML private TextField password;
     @FXML private TextField javaHome;
+    @FXML private TextField jvmArgs;
     @FXML private TextField deployDir;
     @FXML private ProgressIndicator progress;
+    @FXML private Label command;
     @FXML private Button deployButton;
     @FXML private Label status;
     @FXML private TextArea console;
@@ -59,18 +62,20 @@ public class DeployController {
     @Inject private EventBus eventBus;
     @Inject private Project project;
     @Inject private Pipeline pipeline;
-    @Inject private Logger logger;
 
+    @Inject private Logger logger;
     private final BooleanProperty deploying = new SimpleBooleanProperty(this, "deploying", false);
     private Optional<Thread> deployThread = Optional.empty();
 
     @FXML
     public void initialize() {
-        loadSettings(pipeline.getProjectSettings());
-
         deployButton.disableProperty().bind(deploying);
         progress.disableProperty().bind(deploying.not());
         deploying.addListener((o, b, d) -> progress.setProgress(d ? ProgressIndicator.INDETERMINATE_PROGRESS : 0));
+        command.textProperty().bind(Bindings.concat(javaHome.textProperty(), "/bin/java ", jvmArgs.textProperty(),
+                " -jar '", deployDir.textProperty(), "/", GRIP_JAR, "' '", deployDir.textProperty(), "/", PROJECT_FILE, "'"));
+
+        loadSettings(pipeline.getProjectSettings());
     }
 
     @Subscribe
@@ -85,6 +90,7 @@ public class DeployController {
         address.setText(settings.getDeployAddress());
         user.setText(settings.getDeployUser());
         javaHome.setText(settings.getDeployJavaHome());
+        jvmArgs.setText(settings.getDeployJvmOptions());
         deployDir.setText(settings.getDeployDir());
     }
 
@@ -95,6 +101,7 @@ public class DeployController {
         settings.setDeployAddress(address.getText());
         settings.setDeployUser(user.getText());
         settings.setDeployJavaHome(javaHome.getText());
+        settings.setDeployJvmOptions(jvmArgs.getText());
         settings.setDeployDir(deployDir.getText());
         eventBus.post(new ProjectSettingsChangedEvent(settings));
     }
@@ -107,8 +114,7 @@ public class DeployController {
         console.clear();
 
         // Start the deploy in a new thread, so the GUI doesn't freeze
-        deployThread = Optional.of(new Thread(() ->
-                deploy(address.getText(), user.getText(), password.getText(), javaHome.getText(), deployDir.getText())));
+        deployThread = Optional.of(new Thread(this::deploy));
         deployThread.get().setDaemon(true);
         deployThread.get().start();
     }
@@ -124,13 +130,13 @@ public class DeployController {
      * Upload and run the GRIP project using the current deploy settings.  This is run in a separate thread, and it
      * periodically updates the GUI to inform the user of the current status of the deployment.
      */
-    private void deploy(String address, String user, String password, String javaHome, String deployDir) {
-        setStatusAsync("Connecting to " + address, false);
+    private void deploy() {
+        setStatusAsync("Connecting to " + address.getText(), false);
 
         try (SSHClient ssh = new SSHClient()) {
             ssh.addHostKeyVerifier((hostname, port, key) -> true);
-            ssh.connect(address);
-            ssh.authPassword(user, password);
+            ssh.connect(address.getText());
+            ssh.authPassword(user.getText(), password.getText());
 
             // Update the progress bar and status text while uploading files
             SCPFileTransfer scp = ssh.newSCPFileTransfer();
@@ -151,8 +157,8 @@ public class DeployController {
             project.save(projectWriter);
 
             // Upload the GRIP core JAR and the serialized project to the robot
-            scp.upload(new StringInMemoryFile(PROJECT_FILE, projectWriter.toString()), deployDir + "/");
-            scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), deployDir + "/" + GRIP_JAR);
+            scp.upload(new StringInMemoryFile(PROJECT_FILE, projectWriter.toString()), deployDir.getText() + "/");
+            scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), deployDir.getText() + "/" + GRIP_JAR);
 
             // Stop the pipeline before running it remotely, so the two instances of GRIP don't try to publish to the
             // same NetworkTables keys.
@@ -162,8 +168,9 @@ public class DeployController {
             setStatusAsync("Running GRIP", false);
             Session session = ssh.startSession();
             session.allocateDefaultPTY();
-            Session.Command cmd = session.exec(javaHome + "/bin/java -jar " + deployDir + "/" + GRIP_JAR + "' '"
-                    + deployDir + "/" + PROJECT_FILE + "'");
+
+            logger.info("Executing " + command.getText());
+            Session.Command cmd = session.exec(command.getText());
 
             LineReader inputReader = new LineReader(new InputStreamReader(cmd.getInputStream()));
             while (isNotCanceled()) {
@@ -175,7 +182,7 @@ public class DeployController {
                 Platform.runLater(() -> console.setText(console.getText() + line + "\n"));
             }
         } catch (UnknownHostException e) {
-            setStatusAsync("Unknown host: " + address, true);
+            setStatusAsync("Unknown host: " + address.getText(), true);
         } catch (UserAuthException e) {
             setStatusAsync("Invalid username or password (should be \"lvuser\" and blank for roboRIO)", true);
         } catch (InterruptedIOException e) {

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -115,7 +115,7 @@ public class DeployController {
         console.clear();
 
         // Start the deploy in a new thread, so the GUI doesn't freeze
-        deployThread = Optional.of(new Thread(this::deploy));
+        deployThread = Optional.of(new Thread(this::deploy, "Deploy"));
         deployThread.get().setDaemon(true);
         deployThread.get().start();
     }

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -47,7 +47,6 @@ import java.util.logging.Logger;
 public class DeployController {
     private final static String GRIP_JAR = "grip.jar";
     private final static String GRIP_WRAPPER = "grip";
-    private final static String PROJECT_FILE = "project.grip";
     private final static String LOCAL_GRIP_JAR = URLDecoder.decode(
             Project.class.getProtectionDomain().getCodeSource().getLocation().getPath());
 
@@ -57,6 +56,7 @@ public class DeployController {
     @FXML private TextField javaHome;
     @FXML private TextField jvmArgs;
     @FXML private TextField deployDir;
+    @FXML private TextField projectFile;
     @FXML private ProgressIndicator progress;
     @FXML private Label status;
     @FXML private TextArea console;
@@ -74,7 +74,7 @@ public class DeployController {
     public void initialize() {
         deploying.addListener((o, b, d) -> progress.setProgress(d ? ProgressIndicator.INDETERMINATE_PROGRESS : 0));
         command.bind(Bindings.concat(javaHome.textProperty(), "/bin/java ", jvmArgs.textProperty(), " -jar '",
-                deployDir.textProperty(), "/", GRIP_JAR, "' '", deployDir.textProperty(), "/", PROJECT_FILE, "'"));
+                deployDir.textProperty(), "/", GRIP_JAR, "' '", deployDir.textProperty(), "/", projectFile.textProperty(), "'"));
 
         loadSettings(pipeline.getProjectSettings());
     }
@@ -162,7 +162,7 @@ public class DeployController {
             final String pathStr = deployDir.getText() + "/";
             scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), pathStr + GRIP_JAR);
             scp.upload(new StringInMemoryFile(GRIP_WRAPPER, "echo \"" + commandStr + "\"\n" + commandStr, 0755), pathStr);
-            scp.upload(new StringInMemoryFile(PROJECT_FILE, projectWriter.toString()), pathStr);
+            scp.upload(new StringInMemoryFile(projectFile.getText(), projectWriter.toString()), pathStr);
 
             // Stop the pipeline before running it remotely, so the two instances of GRIP don't try to publish to the
             // same NetworkTables keys.

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -2,7 +2,9 @@ package edu.wpi.grip.ui;
 
 import com.google.common.eventbus.EventBus;
 import com.google.common.eventbus.Subscribe;
+import com.google.common.hash.Hashing;
 import com.google.common.io.LineReader;
+import com.google.common.io.Resources;
 import edu.wpi.grip.core.Pipeline;
 import edu.wpi.grip.core.events.ProjectSettingsChangedEvent;
 import edu.wpi.grip.core.events.StopPipelineEvent;
@@ -27,10 +29,8 @@ import net.schmizz.sshj.xfer.LoggingTransferListener;
 import net.schmizz.sshj.xfer.scp.SCPFileTransfer;
 
 import javax.inject.Inject;
-import java.io.IOException;
-import java.io.InputStreamReader;
-import java.io.InterruptedIOException;
-import java.io.StringWriter;
+import java.io.*;
+import java.net.URL;
 import java.net.URLDecoder;
 import java.net.UnknownHostException;
 import java.util.Optional;
@@ -47,8 +47,8 @@ import java.util.logging.Logger;
 public class DeployController {
     private final static String GRIP_JAR = "grip.jar";
     private final static String GRIP_WRAPPER = "grip";
-    private final static String LOCAL_GRIP_JAR = URLDecoder.decode(
-            Project.class.getProtectionDomain().getCodeSource().getLocation().getPath());
+    private final static URL LOCAL_GRIP_URL = Project.class.getProtectionDomain().getCodeSource().getLocation();
+    private final static String LOCAL_GRIP_PATH = URLDecoder.decode(LOCAL_GRIP_URL.getPath());
 
     @FXML private TextField address;
     @FXML private TextField user;
@@ -157,10 +157,29 @@ public class DeployController {
             StringWriter projectWriter = new StringWriter();
             project.save(projectWriter);
 
-            // Upload the GRIP core JAR, a wrapper script to run it, and the serialized project to the robot
             final String commandStr = command.get();
             final String pathStr = deployDir.getText() + "/";
-            scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), pathStr + GRIP_JAR);
+
+
+            // Upload the core GRIP JAR only if there isn't already one with the same hash on the robot.  This prevents
+            // us from redundantly deploying the same JAR over and over again (so deploy times after the first are
+            // much faster), while still ensuring that the JAR is deployed if it has to be.
+            try (Session session = ssh.startSession()) {
+                Session.Command md5Cmd = session.exec("md5sum " + pathStr + GRIP_JAR);
+                String remoteMd5Sum = new DataInputStream(md5Cmd.getInputStream()).readLine();
+                String localMd5Sum = Resources.asByteSource(LOCAL_GRIP_URL).hash(Hashing.md5()).toString();
+
+                // If the MD5 sum doesn't match up or there's any kind of error (there isn't a JAR there, etc...),
+                // just upload the JAR.
+                if (remoteMd5Sum == null || remoteMd5Sum.length() < 32 || !remoteMd5Sum.substring(0, 32).equals(localMd5Sum)) {
+                    scp.upload(new FileSystemFile(LOCAL_GRIP_PATH), pathStr + GRIP_JAR);
+                } else {
+                    logger.info("md5sum " + GRIP_JAR + " matches. Skipping upload.");
+                }
+            }
+
+            // Upload the project file and a wrapper script with the JVM arguments.  These are very small and change
+            // often, so we might as well upload them every time.
             scp.upload(new StringInMemoryFile(GRIP_WRAPPER, "echo \"" + commandStr + "\"\n" + commandStr, 0755), pathStr);
             scp.upload(new StringInMemoryFile(projectFile.getText(), projectWriter.toString()), pathStr);
 
@@ -170,10 +189,10 @@ public class DeployController {
 
             // Run the project!
             setStatusAsync("Running GRIP", false);
+            logger.info("Running " + commandStr);
+
             Session session = ssh.startSession();
             session.allocateDefaultPTY();
-
-            logger.info("Running " + commandStr);
             Session.Command cmd = session.exec(String.format("'%s/%s'", pathStr, GRIP_WRAPPER));
 
             LineReader inputReader = new LineReader(new InputStreamReader(cmd.getInputStream()));

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -13,7 +13,10 @@ import javafx.application.Platform;
 import javafx.beans.binding.Bindings;
 import javafx.beans.property.BooleanProperty;
 import javafx.fxml.FXML;
-import javafx.scene.control.*;
+import javafx.scene.control.Label;
+import javafx.scene.control.ProgressIndicator;
+import javafx.scene.control.TextArea;
+import javafx.scene.control.TextField;
 import net.schmizz.sshj.SSHClient;
 import net.schmizz.sshj.common.StreamCopier;
 import net.schmizz.sshj.connection.channel.direct.Session;
@@ -42,6 +45,7 @@ import java.util.logging.Logger;
  */
 public class DeployController {
     private final static String GRIP_JAR = "grip.jar";
+    private final static String GRIP_WRAPPER = "grip";
     private final static String PROJECT_FILE = "project.grip";
     private final static String LOCAL_GRIP_JAR = URLDecoder.decode(
             Project.class.getProtectionDomain().getCodeSource().getLocation().getPath());
@@ -152,9 +156,12 @@ public class DeployController {
             StringWriter projectWriter = new StringWriter();
             project.save(projectWriter);
 
-            // Upload the GRIP core JAR and the serialized project to the robot
-            scp.upload(new StringInMemoryFile(PROJECT_FILE, projectWriter.toString()), deployDir.getText() + "/");
-            scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), deployDir.getText() + "/" + GRIP_JAR);
+            // Upload the GRIP core JAR, a wrapper script to run it, and the serialized project to the robot
+            final String commandStr = command.getText();
+            final String pathStr = deployDir.getText() + "/";
+            scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), pathStr + GRIP_JAR);
+            scp.upload(new StringInMemoryFile(GRIP_WRAPPER, "echo \"" + commandStr + "\"\n" + commandStr, 0755), pathStr);
+            scp.upload(new StringInMemoryFile(PROJECT_FILE, projectWriter.toString()), pathStr);
 
             // Stop the pipeline before running it remotely, so the two instances of GRIP don't try to publish to the
             // same NetworkTables keys.
@@ -165,8 +172,7 @@ public class DeployController {
             Session session = ssh.startSession();
             session.allocateDefaultPTY();
 
-            logger.info("Executing " + command.getText());
-            Session.Command cmd = session.exec(command.getText());
+            Session.Command cmd = session.exec(String.format("'%s/%s'", pathStr, GRIP_WRAPPER));
 
             LineReader inputReader = new LineReader(new InputStreamReader(cmd.getInputStream()));
             while (isNotCanceled()) {

--- a/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/DeployController.java
@@ -12,6 +12,7 @@ import edu.wpi.grip.ui.util.StringInMemoryFile;
 import javafx.application.Platform;
 import javafx.beans.binding.Bindings;
 import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.StringProperty;
 import javafx.fxml.FXML;
 import javafx.scene.control.Label;
 import javafx.scene.control.ProgressIndicator;
@@ -57,10 +58,10 @@ public class DeployController {
     @FXML private TextField jvmArgs;
     @FXML private TextField deployDir;
     @FXML private ProgressIndicator progress;
-    @FXML private Label command;
     @FXML private Label status;
     @FXML private TextArea console;
     @FXML private BooleanProperty deploying;
+    @FXML private StringProperty command;
 
     @Inject private EventBus eventBus;
     @Inject private Project project;
@@ -72,8 +73,8 @@ public class DeployController {
     @FXML
     public void initialize() {
         deploying.addListener((o, b, d) -> progress.setProgress(d ? ProgressIndicator.INDETERMINATE_PROGRESS : 0));
-        command.textProperty().bind(Bindings.concat(javaHome.textProperty(), "/bin/java ", jvmArgs.textProperty(),
-                " -jar '", deployDir.textProperty(), "/", GRIP_JAR, "' '", deployDir.textProperty(), "/", PROJECT_FILE, "'"));
+        command.bind(Bindings.concat(javaHome.textProperty(), "/bin/java ", jvmArgs.textProperty(), " -jar '",
+                deployDir.textProperty(), "/", GRIP_JAR, "' '", deployDir.textProperty(), "/", PROJECT_FILE, "'"));
 
         loadSettings(pipeline.getProjectSettings());
     }
@@ -157,7 +158,7 @@ public class DeployController {
             project.save(projectWriter);
 
             // Upload the GRIP core JAR, a wrapper script to run it, and the serialized project to the robot
-            final String commandStr = command.getText();
+            final String commandStr = command.get();
             final String pathStr = deployDir.getText() + "/";
             scp.upload(new FileSystemFile(LOCAL_GRIP_JAR), pathStr + GRIP_JAR);
             scp.upload(new StringInMemoryFile(GRIP_WRAPPER, "echo \"" + commandStr + "\"\n" + commandStr, 0755), pathStr);
@@ -172,6 +173,7 @@ public class DeployController {
             Session session = ssh.startSession();
             session.allocateDefaultPTY();
 
+            logger.info("Running " + commandStr);
             Session.Command cmd = session.exec(String.format("'%s/%s'", pathStr, GRIP_WRAPPER));
 
             LineReader inputReader = new LineReader(new InputStreamReader(cmd.getInputStream()));

--- a/ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java
@@ -196,6 +196,8 @@ public class MainWindowController {
         graphic.setFitWidth(DPIUtility.SMALL_ICON_SIZE);
         graphic.setFitHeight(DPIUtility.SMALL_ICON_SIZE);
 
+        deployPane.requestFocus();
+
         Dialog<ButtonType> dialog = new Dialog<>();
         dialog.setTitle("Deploy");
         dialog.setHeaderText("Deploy");

--- a/ui/src/main/java/edu/wpi/grip/ui/util/StringInMemoryFile.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/util/StringInMemoryFile.java
@@ -12,11 +12,17 @@ import java.io.StringBufferInputStream;
  */
 public class StringInMemoryFile extends InMemorySourceFile {
     private final String name, contents;
+    private final int permissions;
 
-    public StringInMemoryFile(String name, String contents) {
+    public StringInMemoryFile(String name, String contents, int permissions) {
         super();
         this.name = name;
         this.contents = contents;
+        this.permissions = permissions;
+    }
+
+    public StringInMemoryFile(String name, String contents) {
+        this(name, contents, 0644);
     }
 
     @Override
@@ -27,6 +33,11 @@ public class StringInMemoryFile extends InMemorySourceFile {
     @Override
     public long getLength() {
         return contents.getBytes().length;
+    }
+
+    @Override
+    public int getPermissions() {
+        return permissions;
     }
 
     @Override

--- a/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
+++ b/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
@@ -11,6 +11,10 @@
 <VBox styleClass="deploy-pane" maxWidth="Infinity" xmlns:fx="http://javafx.com/fxml/1"
       onKeyPressed="if (event.code == javafx.scene.input.KeyCode.ENTER) { controller.onDeploy() }"
       xmlns="http://javafx.com/javafx/null" fx:controller="edu.wpi.grip.ui.DeployController">
+    <fx:define>
+        <SimpleBooleanProperty fx:id="deploying"/>
+    </fx:define>
+
     <GridPane maxHeight="Infinity">
         <columnConstraints>
             <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
@@ -59,6 +63,7 @@
         <buttons>
             <Button fx:id="deployButton" defaultButton="true" onMouseClicked="#onDeploy" text="Deploy"
                     ButtonBar.buttonData="APPLY">
+                <fx:script>deployButton.disableProperty().bind(deploying)</fx:script>
                 <graphic>
                     <ImageView styleClass="menu-graphic">
                         <fitWidth>
@@ -78,7 +83,9 @@
     </ButtonBar>
 
     <StackPane>
-        <ProgressBar fx:id="progress" maxWidth="Infinity" progress="0"/>
+        <ProgressBar fx:id="progress" maxWidth="Infinity" progress="0">
+            <fx:script>progress.disableProperty().bind(deploying.not())</fx:script>
+        </ProgressBar>
         <Label fx:id="status"/>
     </StackPane>
 

--- a/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
+++ b/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
@@ -2,6 +2,7 @@
 
 <?language javascript?>
 
+<?import edu.wpi.grip.ui.DeployController?>
 <?import edu.wpi.grip.ui.util.DPIUtility?>
 <?import javafx.beans.property.SimpleBooleanProperty?>
 <?import javafx.beans.property.SimpleStringProperty?>
@@ -36,6 +37,10 @@
             <PasswordField disable="${deployButton.disabled}" fx:id="password" promptText="Password"
                            GridPane.columnIndex="1" GridPane.rowIndex="2"/>
 
+            <Label disable="${deployButton.disabled}" text="Project File" GridPane.columnIndex="0"
+                   GridPane.rowIndex="3"/>
+            <TextField disable="${deployButton.disabled}" fx:id="projectFile" promptText="Project File"
+                       GridPane.columnIndex="1" GridPane.rowIndex="3" text="project.grip"/>
         </GridPane>
         <Separator orientation="VERTICAL" HBox.hgrow="NEVER" visible="false">
             <prefWidth>
@@ -61,33 +66,34 @@
                    GridPane.rowIndex="2"/>
             <TextField disable="${deployButton.disabled}" fx:id="jvmArgs" promptText="JVM Arguments"
                        GridPane.columnIndex="1" GridPane.rowIndex="2"/>
+
+            <ButtonBar GridPane.columnIndex="0" GridPane.rowIndex="3" GridPane.columnSpan="2">
+                <buttons>
+                    <Button fx:id="deployButton" defaultButton="true" onMouseClicked="#onDeploy" text="Deploy"
+                            ButtonBar.buttonData="APPLY">
+                        <fx:script>deployButton.disableProperty().bind(deploying)</fx:script>
+                        <graphic>
+                            <ImageView styleClass="menu-graphic">
+                                <fitWidth>
+                                    <DPIUtility fx:constant="SMALL_ICON_SIZE"/>
+                                </fitWidth>
+                                <fitHeight>
+                                    <DPIUtility fx:constant="SMALL_ICON_SIZE"/>
+                                </fitHeight>
+                                <image>
+                                    <Image url="@icons/deploy.png"/>
+                                </image>
+                            </ImageView>
+                        </graphic>
+                    </Button>
+                    <Button text="Stop" ButtonBar.buttonData="CANCEL_CLOSE" onMouseClicked="#onStop"/>
+                </buttons>
+            </ButtonBar>
+
         </GridPane>
     </HBox>
 
     <Separator orientation="HORIZONTAL" VBox.vgrow="NEVER"/>
-
-    <ButtonBar>
-        <buttons>
-            <Button fx:id="deployButton" defaultButton="true" onMouseClicked="#onDeploy" text="Deploy"
-                    ButtonBar.buttonData="APPLY">
-                <fx:script>deployButton.disableProperty().bind(deploying)</fx:script>
-                <graphic>
-                    <ImageView styleClass="menu-graphic">
-                        <fitWidth>
-                            <DPIUtility fx:constant="SMALL_ICON_SIZE"/>
-                        </fitWidth>
-                        <fitHeight>
-                            <DPIUtility fx:constant="SMALL_ICON_SIZE"/>
-                        </fitHeight>
-                        <image>
-                            <Image url="@icons/deploy.png"/>
-                        </image>
-                    </ImageView>
-                </graphic>
-            </Button>
-            <Button text="Stop" ButtonBar.buttonData="CANCEL_CLOSE" onMouseClicked="#onStop"/>
-        </buttons>
-    </ButtonBar>
 
     <StackPane>
         <ProgressBar fx:id="progress" maxWidth="Infinity" progress="0">

--- a/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
+++ b/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
@@ -1,18 +1,26 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
-<?language javascript ?>
+<?language javascript?>
 
 <?import edu.wpi.grip.ui.util.DPIUtility?>
-<?import edu.wpi.grip.ui.DeployController?>
+<?import javafx.beans.property.SimpleBooleanProperty?>
 <?import javafx.scene.control.*?>
 <?import javafx.scene.image.Image?>
 <?import javafx.scene.image.ImageView?>
 <?import javafx.scene.layout.*?>
 <VBox styleClass="deploy-pane" maxWidth="Infinity" xmlns:fx="http://javafx.com/fxml/1"
+      onKeyPressed="if (event.code == javafx.scene.input.KeyCode.ENTER) { controller.onDeploy() }"
       xmlns="http://javafx.com/javafx/null" fx:controller="edu.wpi.grip.ui.DeployController">
     <GridPane maxHeight="Infinity">
         <columnConstraints>
-            <ColumnConstraints hgrow="NEVER"/>
+            <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
+            <ColumnConstraints hgrow="ALWAYS"/>
+            <ColumnConstraints hgrow="NEVER">
+                <prefWidth>
+                    <DPIUtility fx:constant="FONT_SIZE"/>
+                </prefWidth>
+            </ColumnConstraints>
+            <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
             <ColumnConstraints hgrow="ALWAYS"/>
         </columnConstraints>
 
@@ -21,22 +29,31 @@
                    GridPane.rowIndex="0"/>
 
         <Label disable="${deployButton.disabled}" text="User" GridPane.columnIndex="0" GridPane.rowIndex="1"/>
-        <TextField disable="${deployButton.disabled}" fx:id="user" promptText="User" GridPane.columnIndex="1"
+        <TextField disable="${deployButton.disabled}" fx:id="user" promptText="Username" GridPane.columnIndex="1"
                    GridPane.rowIndex="1"/>
 
         <Label disable="${deployButton.disabled}" text="Password" GridPane.columnIndex="0" GridPane.rowIndex="2"/>
         <PasswordField disable="${deployButton.disabled}" fx:id="password" promptText="Password"
                        GridPane.columnIndex="1" GridPane.rowIndex="2"/>
 
-        <Label disable="${deployButton.disabled}" text="Java Home" GridPane.columnIndex="0" GridPane.rowIndex="3"/>
-        <TextField disable="${deployButton.disabled}" fx:id="javaHome" promptText="Java Home"
-                   GridPane.columnIndex="1" GridPane.rowIndex="3"/>
-
-        <Label disable="${deployButton.disabled}" text="Deploy Directory" GridPane.columnIndex="0"
-               GridPane.rowIndex="4"/>
+        <Label disable="${deployButton.disabled}" text="Deploy Directory" GridPane.columnIndex="3"
+               GridPane.rowIndex="0"/>
         <TextField disable="${deployButton.disabled}" fx:id="deployDir" promptText="Deploy Directory"
-                   GridPane.columnIndex="1" GridPane.rowIndex="4"/>
+                   GridPane.columnIndex="4" GridPane.rowIndex="0"/>
+
+        <Label disable="${deployButton.disabled}" text="Java Home" GridPane.columnIndex="3" GridPane.rowIndex="1"/>
+        <TextField disable="${deployButton.disabled}" fx:id="javaHome" promptText="Java Home"
+                   GridPane.columnIndex="4" GridPane.rowIndex="1"/>
+
+        <Label disable="${deployButton.disabled}" text="JVM Arguments" GridPane.columnIndex="3" GridPane.rowIndex="2"/>
+        <TextField disable="${deployButton.disabled}" fx:id="jvmArgs" promptText="JVM Arguments"
+                   GridPane.columnIndex="4" GridPane.rowIndex="2"/>
+
+        <Label disable="${deployButton.disabled}" text="Command" GridPane.columnIndex="0" GridPane.rowIndex="3"/>
+        <Label fx:id="command" disable="${deployButton.disabled}" GridPane.columnIndex="1" GridPane.columnSpan="4"
+               GridPane.rowIndex="3" style="-fx-font-family: monospace"/>
     </GridPane>
+
 
     <ButtonBar>
         <buttons>
@@ -65,5 +82,5 @@
         <Label fx:id="status"/>
     </StackPane>
 
-    <TextArea fx:id="console" editable="false" styleClass="console" prefRowCount="24" prefColumnCount="80"/>
+    <TextArea fx:id="console" editable="false" styleClass="console" prefRowCount="24" prefColumnCount="100"/>
 </VBox>

--- a/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
+++ b/ui/src/main/resources/edu/wpi/grip/ui/Deploy.fxml
@@ -4,60 +4,67 @@
 
 <?import edu.wpi.grip.ui.util.DPIUtility?>
 <?import javafx.beans.property.SimpleBooleanProperty?>
+<?import javafx.beans.property.SimpleStringProperty?>
 <?import javafx.scene.control.*?>
 <?import javafx.scene.image.Image?>
 <?import javafx.scene.image.ImageView?>
 <?import javafx.scene.layout.*?>
 <VBox styleClass="deploy-pane" maxWidth="Infinity" xmlns:fx="http://javafx.com/fxml/1"
       onKeyPressed="if (event.code == javafx.scene.input.KeyCode.ENTER) { controller.onDeploy() }"
-      xmlns="http://javafx.com/javafx/null" fx:controller="edu.wpi.grip.ui.DeployController">
+      xmlns="http://javafx.com/javafx/null" fx:controller="edu.wpi.grip.ui.DeployController" fillWidth="true">
     <fx:define>
         <SimpleBooleanProperty fx:id="deploying"/>
+        <SimpleStringProperty fx:id="command"/>
     </fx:define>
 
-    <GridPane maxHeight="Infinity">
-        <columnConstraints>
-            <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
-            <ColumnConstraints hgrow="ALWAYS"/>
-            <ColumnConstraints hgrow="NEVER">
-                <prefWidth>
-                    <DPIUtility fx:constant="FONT_SIZE"/>
-                </prefWidth>
-            </ColumnConstraints>
-            <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
-            <ColumnConstraints hgrow="ALWAYS"/>
-        </columnConstraints>
+    <HBox maxWidth="Infinity">
+        <GridPane maxHeight="Infinity" HBox.hgrow="ALWAYS">
+            <columnConstraints>
+                <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
+                <ColumnConstraints hgrow="ALWAYS"/>
+            </columnConstraints>
 
-        <Label disable="${deployButton.disabled}" text="Address" GridPane.columnIndex="0" GridPane.rowIndex="0"/>
-        <TextField disable="${deployButton.disabled}" fx:id="address" promptText="Address" GridPane.columnIndex="1"
+            <Label disable="${deployButton.disabled}" text="Address" GridPane.columnIndex="0" GridPane.rowIndex="0"/>
+            <TextField disable="${deployButton.disabled}" fx:id="address" promptText="Address" GridPane.columnIndex="1"
+                       GridPane.rowIndex="0"/>
+
+            <Label disable="${deployButton.disabled}" text="User" GridPane.columnIndex="0" GridPane.rowIndex="1"/>
+            <TextField disable="${deployButton.disabled}" fx:id="user" promptText="Username" GridPane.columnIndex="1"
+                       GridPane.rowIndex="1"/>
+
+            <Label disable="${deployButton.disabled}" text="Password" GridPane.columnIndex="0" GridPane.rowIndex="2"/>
+            <PasswordField disable="${deployButton.disabled}" fx:id="password" promptText="Password"
+                           GridPane.columnIndex="1" GridPane.rowIndex="2"/>
+
+        </GridPane>
+        <Separator orientation="VERTICAL" HBox.hgrow="NEVER" visible="false">
+            <prefWidth>
+                <DPIUtility fx:constant="SMALL_ICON_SIZE"/>
+            </prefWidth>
+        </Separator>
+        <GridPane maxHeight="Infinity" HBox.hgrow="ALWAYS">
+            <columnConstraints>
+                <ColumnConstraints hgrow="NEVER" halignment="RIGHT"/>
+                <ColumnConstraints hgrow="ALWAYS"/>
+            </columnConstraints>
+
+            <Label disable="${deployButton.disabled}" text="Deploy Directory" GridPane.columnIndex="0"
                    GridPane.rowIndex="0"/>
+            <TextField disable="${deployButton.disabled}" fx:id="deployDir" promptText="Deploy Directory"
+                       GridPane.columnIndex="1" GridPane.rowIndex="0"/>
 
-        <Label disable="${deployButton.disabled}" text="User" GridPane.columnIndex="0" GridPane.rowIndex="1"/>
-        <TextField disable="${deployButton.disabled}" fx:id="user" promptText="Username" GridPane.columnIndex="1"
-                   GridPane.rowIndex="1"/>
+            <Label disable="${deployButton.disabled}" text="Java Home" GridPane.columnIndex="0" GridPane.rowIndex="1"/>
+            <TextField disable="${deployButton.disabled}" fx:id="javaHome" promptText="Java Home"
+                       GridPane.columnIndex="1" GridPane.rowIndex="1"/>
 
-        <Label disable="${deployButton.disabled}" text="Password" GridPane.columnIndex="0" GridPane.rowIndex="2"/>
-        <PasswordField disable="${deployButton.disabled}" fx:id="password" promptText="Password"
+            <Label disable="${deployButton.disabled}" text="JVM Arguments" GridPane.columnIndex="0"
+                   GridPane.rowIndex="2"/>
+            <TextField disable="${deployButton.disabled}" fx:id="jvmArgs" promptText="JVM Arguments"
                        GridPane.columnIndex="1" GridPane.rowIndex="2"/>
+        </GridPane>
+    </HBox>
 
-        <Label disable="${deployButton.disabled}" text="Deploy Directory" GridPane.columnIndex="3"
-               GridPane.rowIndex="0"/>
-        <TextField disable="${deployButton.disabled}" fx:id="deployDir" promptText="Deploy Directory"
-                   GridPane.columnIndex="4" GridPane.rowIndex="0"/>
-
-        <Label disable="${deployButton.disabled}" text="Java Home" GridPane.columnIndex="3" GridPane.rowIndex="1"/>
-        <TextField disable="${deployButton.disabled}" fx:id="javaHome" promptText="Java Home"
-                   GridPane.columnIndex="4" GridPane.rowIndex="1"/>
-
-        <Label disable="${deployButton.disabled}" text="JVM Arguments" GridPane.columnIndex="3" GridPane.rowIndex="2"/>
-        <TextField disable="${deployButton.disabled}" fx:id="jvmArgs" promptText="JVM Arguments"
-                   GridPane.columnIndex="4" GridPane.rowIndex="2"/>
-
-        <Label disable="${deployButton.disabled}" text="Command" GridPane.columnIndex="0" GridPane.rowIndex="3"/>
-        <Label fx:id="command" disable="${deployButton.disabled}" GridPane.columnIndex="1" GridPane.columnSpan="4"
-               GridPane.rowIndex="3" style="-fx-font-family: monospace"/>
-    </GridPane>
-
+    <Separator orientation="HORIZONTAL" VBox.vgrow="NEVER"/>
 
     <ButtonBar>
         <buttons>
@@ -89,5 +96,6 @@
         <Label fx:id="status"/>
     </StackPane>
 
-    <TextArea fx:id="console" editable="false" styleClass="console" prefRowCount="24" prefColumnCount="100"/>
+    <TextArea fx:id="console" editable="false" styleClass="console" prefRowCount="24" prefColumnCount="100"
+              VBox.vgrow="ALWAYS"/>
 </VBox>

--- a/ui/src/main/resources/edu/wpi/grip/ui/GRIP.css
+++ b/ui/src/main/resources/edu/wpi/grip/ui/GRIP.css
@@ -189,7 +189,7 @@ VBox.sockets {
     -fx-background-color: grey;
 }
 
-.property-sheet .property-pane, .deploy-pane GridPane {
+.property-sheet .property-pane {
     -fx-hgap: 4em;
     -fx-vgap: 1em;
     -fx-padding: 0;
@@ -205,6 +205,12 @@ VBox.sockets {
 
 .deploy-pane {
     -fx-spacing: 1em;
+}
+
+.deploy-pane GridPane {
+    -fx-hgap: 1em;
+    -fx-vgap: 1em;
+    -fx-padding: 0;
 }
 
 .deploy-pane StackPane {

